Mathlib/Analysis/NormedSpace/ProdLp.lean

```diff
@@ -0,0 +1,772 @@
+
+open Real Set Filter IsROrC Bornology BigOperators Uniformity Topology NNReal ENNReal
+
+noncomputable section
+
+variable (p : ℝ≥0∞) (𝕜 α β : Type*)
+
+namespace WithLp
+
+section algebra
+
+/- Register simplification lemmas for the applications of `WithLp p (α × β)` elements, as the usual
+lemmas for `Prod` will not trigger. -/
+
+variable {p 𝕜 α β}
+
+variable [Semiring 𝕜] [AddCommGroup α] [AddCommGroup β]
+variable (x y : WithLp p (α × β)) (c : 𝕜)
+
+@[simp]
+theorem zero_fst : (0 : WithLp p (α × β)).fst = 0 :=
+  rfl
+
+@[simp]
+theorem zero_snd : (0 : WithLp p (α × β)).snd = 0 :=
+  rfl
+
+@[simp]
+theorem add_fst : (x + y).fst = x.fst + y.fst :=
+  rfl
+
+@[simp]
+theorem add_snd : (x + y).snd = x.snd + y.snd :=
+  rfl
+
+@[simp]
+theorem sub_fst : (x - y).fst = x.fst - y.fst :=
+  rfl
+
+@[simp]
+theorem sub_snd : (x - y).snd = x.snd - y.snd :=
+  rfl
+
+@[simp]
+theorem neg_fst : (-x).fst = -x.fst :=
+  rfl
+
+@[simp]
+theorem neg_snd : (-x).snd = -x.snd :=
+  rfl
+
+variable [Module 𝕜 α] [Module 𝕜 β]
+
+@[simp]
+theorem smul_fst : (c • x).fst = c • x.fst :=
+  rfl
+
+@[simp]
+theorem smul_snd : (c • x).snd = c • x.snd :=
+  rfl
+
+end algebra
+
+/-! Note that the unapplied versions of these lemmas are deliberately omitted, as they break
+the use of the type synonym. -/
+
+section equiv
+
+variable {p α β}
+
+@[simp]
+theorem equiv_fst (x : WithLp p (α × β)) : (WithLp.equiv p (α × β) x).fst = x.fst :=
+  rfl
+
+@[simp]
+theorem equiv_snd (x : WithLp p (α × β)) : (WithLp.equiv p (α × β) x).snd = x.snd :=
+  rfl
+
+@[simp]
+theorem equiv_symm_fst (x : α × β) : ((WithLp.equiv p (α × β)).symm x).fst = x.fst :=
+  rfl
+
+@[simp]
+theorem equiv_symm_snd (x : α × β) : ((WithLp.equiv p (α × β)).symm x).snd = x.snd :=
+  rfl
+
+end equiv
+
+section DistNorm
+
+/-!
+### Definition of `edist`, `dist` and `norm` on `WithLp p (α × β)`
+
+In this section we define the `edist`, `dist` and `norm` functions on `WithLp p (α × β)` without
+assuming `[Fact (1 ≤ p)]` or metric properties of the spaces `α` and `β`. This allows us to provide
+the rewrite lemmas for each of three cases `p = 0`, `p = ∞` and `0 < p.toReal`.
+-/
+
+
+section EDist
+
+variable [EDist α] [EDist β]
+
+open Classical in
+/-- Endowing the space `WithLp p (α × β)` with the `L^p` edistance. We register this instance
+separate from `WithLp.instProdPseudoEMetric` since the latter requires the type class hypothesis
+`[Fact (1 ≤ p)]` in order to prove the triangle inequality.
+
+Registering this separately allows for a future emetric-like structure on `WithLp p (α × β)` for
+`p < 1` satisfying a relaxed triangle inequality. The terminology for this varies throughout the
+literature, but it is sometimes called a *quasi-metric* or *semi-metric*. -/
+instance instProdEDist : EDist (WithLp p (α × β)) where
+  edist f g :=
+    if _hp : p = 0 then
+      (if f.fst = g.fst then 0 else 1) + (if f.snd = g.snd then 0 else 1)
```

Comment:
for coherence with the norm, should it be `if edist f.1 g.1 = 0 then 0 else 1 + ...`? (If you prefer, I can do that in a later PR in which I would also change `PiLp`, because I wouldn't want to delay this one even more...)

Reply:
no problem, this is an easy fix.